File: zfs/replicate/filesystem/create.py
"""ZFS FileSystem creation."""

import os.path

from .. import process
from ..command import Command, over_ssh
from ..error import ZFSReplicateError
from ..list import inits
from ..stderr import clean
from . import type
from .list import list
from .type import FileSystem


def create(filesystem: FileSystem, ssh_command: Command) -> None:
    """Create a Remote FileSystem."""
    if filesystem.name is None:
        msg = f"refusing to create dataset: '{filesystem.dataset}'"
        raise ZFSReplicateError(msg, filesystem)

    top_level = type.filesystem(name=filesystem.dataset, readonly=filesystem.readonly)

    filesystems = [x.name for x in list(top_level, ssh_command=ssh_command)]

    for head in inits(filesystem.name.split("/"))[1:]:
        path = os.path.join(*head)

        if path in filesystems:
            continue

        result = process.run(over_ssh(ssh_command, _create(path)))

        if result.returncode:
            error = clean(result.stderr)

            if b"successfully created, but not mounted" in error:
                return  # Ignore this error.

            msg = f"unable to create remote dataset: '{filesystem.dataset}': {error!r}"
            raise ZFSReplicateError(msg, filesystem, error)


def _create(filesystem: str) -> Command:
    return Command.with_empty_env("zfs", "create", "-o", "readonly=on", filesystem)

File: src/FileSystem/Create.hs
module FileSystem.Create where

import           Control                        ( die )
import           Control.Monad                  ( mapM_
                                                , when
                                                )
import           Data.List                      ( inits
                                                , isSubsequenceOf
                                                )
import           FileSystem.Types
import           Prelude                 hiding ( error )
import qualified Shell                          ( process
                                                , sanitize
                                                )
import           System.Exit                    ( ExitCode(ExitSuccess) )
import           System.FilePath                ( joinPath
                                                , splitDirectories
                                                )
import           System.Process                 ( readCreateProcessWithExitCode )

create :: FileSystem -> String -> IO ()
create fs sshCommand = do
  when (null $ name fs) $ die $ "refusing to create dataset: '" ++ dataset fs ++ "'"
  fileSystems <- list topLevel sshCommand
  mapM_ create' $ filter (`notElem` (name <$> fileSystems)) $ ancestors $ name fs
  where topLevel = (fromName $ dataset fs) { readonly = readonly fs }

list :: FileSystem -> String -> IO [FileSystem]
list fs sshCommand = do
  (code, stdout, stderr) <- readCreateProcessWithExitCode (Shell.process command) ""
  let error = Shell.sanitize stderr
  when (code /= ExitSuccess) $ die $ "error encountered while listing filesystems of '" ++ name fs ++ "': " ++ error
  return $ filesystems stdout
 where
  command = sshCommand ++ "/usr/bin/env - zfs list " ++ unwords options ++ " " ++ name fs
  options = ["-H", "-o name,readonly", "-t filesystem,volume", "-r"]

ancestors :: FilePath -> [FilePath]
ancestors = tail . fmap joinPath . inits . splitDirectories

create' :: FilePath -> IO ()
create' path = do
  (code, _stdout, stderr) <- readCreateProcessWithExitCode (Shell.process command) ""
  let error = Shell.sanitize stderr
  when (code /= ExitSuccess && not ("successfully created, but not mounted" `isSubsequenceOf` error))
    $  die
    $  "unable to create remote dataset: "
    ++ path
  where command = "/usr/bin/env - zfs create -o readonly=on " ++ path

filesystems :: String -> [FileSystem]
filesystems = fmap filesystem . filter (not . null) . lines

filesystem :: String -> FileSystem
filesystem input = (fromName n) { readonly = ro == "on" } where (n, '\t' : ro) = span (/= '\t') input
